### PR TITLE
Azure repos: added await & retry logic based on rate limit headers

### PR DESCRIPTION
**Pull request by tovbinm**
## Description

Azure repos: added await & retry logic based on rate limit headers

## Type of change
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change

## Extra info

Assuming Axios response headers are lowercased

### CODE REVIEW

File: sources/azure-repos-source/src/azure-repos.ts [hunk 1]
```diff
@@ -170,6 +179,19 @@ export class AzureRepos {
       const res = this.httpClient.get<T, R>(path, conf);
       return res;
     } catch (err: any) {
+      const retryAfterSecs = err?.response?.headers['retry-after'];
```
Comment (thomas-gerber) [inline sources/azure-repos-source/src/azure-repos.ts:182]: according to the documentation, this header also appears on successful requests; I think it is important to look for it there for proper throttling. Catching it during errors should be a fallback.

File: sources/azure-repos-source/src/azure-repos.ts [hunk 2]
```diff
@@ -170,6 +179,19 @@ export class AzureRepos {
...
+      if (retryAfterSecs) {
```
Comment (thomas-gerber) [inline sources/azure-repos-source/src/azure-repos.ts:183]: when we catch it during errors, do we have a way to stop trying after a while?

**Review by thomas-gerber — CHANGES_REQUESTED** [review]

File: sources/azure-repos-source/src/azure-repos.ts (hunk 1 above)
Comment (tovbinm) [reply sources/azure-repos-source/src/azure-repos.ts:182]: see - https://learn.microsoft.com/en-us/azure/devops/integrate/concepts/rate-limits?view=azure-devops#recommendations

File: sources/azure-repos-source/src/azure-repos.ts (hunk 2 above)
Comment (tovbinm) [reply sources/azure-repos-source/src/azure-repos.ts:183]: I can put a limit of retries.

File: sources/azure-repos-source/src/azure-repos.ts (hunk 1 above)
Comment (thomas-gerber) [reply sources/azure-repos-source/src/azure-repos.ts:182]: yes exactly; here you only check if an error is caught; you should also check if the request is successful and wait there as well.
`We recommend that you at least respond to the Retry-After header. If you detect a Retry-After header in **any** response, wait until that amount of time has passed before sending another request. Doing so helps your client application experience fewer enforced delays. Keep in mind that **the response is 200**, so you don't need to apply retry logic to the request.`

Comment (tovbinm) [reply sources/azure-repos-source/src/azure-repos.ts]: added

Comment (tovbinm) [reply sources/azure-repos-source/src/azure-repos.ts]: fixed

**Review by thomas-gerber — APPROVED** [review]
Approved with the caveat that there is no test for it

**Pull request merged**